### PR TITLE
nix: upgrade nixpkgs to latest nixos-23.05

### PR DESCRIPTION
**Pull request by jakubgs**
Notable upgrades:

* Bash `5.1` to `5.2`
* Git `2.37.3` to `2.40.1`
* Curl `7.85.0` to `8.0.1`
* OpenSSL `3.0.5` to `3.0.8`
* Go `1.18.6` to `1.18.9`
* NodeJS `18.9.1` to `18.16.0`
* Java `1.8.0_322` to `1.8.0_362`
* Ruby `3.1.2` to `3.1.4`
* Python `2.7.18` to `3.10.11`
* Clojure `1.11.1.1165` to `1.11.1.1273`
* Clj-kondo `v2022.10.05` to `v2023.04.14`
* Zprint `1.2.5` to `1.2.6`
* Gradle `6.9.2` to `6.9.4`
* Bundler `2.3.22` to `2.3.26`
* Android Platform Tools `33.0.2` to `33.0.3`

Removals:

* Zprint since the version in `nixpkgs` was newer than in overlay.
* Xcode wrapper definition was removed since my fixes were merged:
  - https://github.com/NixOS/nixpkgs/pull/204278
  - https://github.com/NixOS/nixpkgs/pull/228696

### CODE REVIEW

**Comment by jakubgs** [comment]
Looks like something is broken with the loader for prebuilt `make` in Android NDK:
```
[nix-shell:~/work/status-mobile]$ $ANDROID_HOME/ndk-bundle/prebuilt/linux-x86_64/bin/make
bash: /nix/store/25lmcjm95fp3k3pvywwk7b6nvs0fpl6i-androidsdk-mod-sdk/ndk-bundle/prebuilt/linux-x86_64/bin/make: No such file or directory

[nix-shell:~/work/status-mobile]$ ldd $ANDROID_HOME/ndk-bundle/prebuilt/linux-x86_64/bin/make
	linux-vdso.so.1 (0x00007fff4e0eb000)
	libdl.so.2 => /nix/store/yzjgl0h6a3qh1mby405428f16xww37h0-glibc-2.35-224/lib/libdl.so.2 (0x00007fbff3827000)
	libc.so.6 => /nix/store/yzjgl0h6a3qh1mby405428f16xww37h0-glibc-2.35-224/lib/libc.so.6 (0x00007fbff361e000)
	/lib64/ld-linux-x86-64.so.2 => /nix/store/yzjgl0h6a3qh1mby405428f16xww37h0-glibc-2.35-224/lib64/ld-linux-x86-64.so.2 (0x00007fbff382e000)
```

**Comment by jakubgs** [comment]
This issue was introduced by this PR and commit due to a typo:

* https://github.com/NixOS/nixpkgs/pull/195752 - https://github.com/NixOS/nixpkgs/commit/a5f43691

I've fixed it in:

* https://github.com/NixOS/nixpkgs/pull/215492

**Comment by flexsurfer** [comment]
still relevant?

**Comment by jakubgs** [comment]
It will be when the React Native upgrade goes through.

**Comment by flexsurfer** [comment]
@jakubgs is there the specific version we're waiting for? i believe one major upgrade has been merged already.

**Comment by jakubgs** [comment]
Actually, now that we don't depend on Gradle 5 I could update this and get it merged:
https://github.com/status-im/status-mobile/blob/56dbb77ee5089441482f3d099ba9c119eb88daf7/nix/overlay.nix#L60
Since just yesterday a new NixOS release was cut:

* https://github.com/NixOS/nixpkgs/issues/223562

Which no longer contains Gradle 5:
https://github.com/NixOS/nixpkgs/blob/3a70dd92993182f8e514700ccf5b1ae9fc8a3b8d/pkgs/development/tools/build-managers/gradle/default.nix#L129-L148

So we can try using that now. Although I'm wondering if maybe I should take this opportunity and teach @yakimant about upgrade of `nixpkgs` in mobile.

**Comment by jakubgs** [comment]
Looks like linting is failing with:
```
src/quo2/components/inputs/title_input/view.cljs:43:38: error: fn is called with 1 arg but expects 0
src/quo2/components/inputs/title_input/view.cljs:44:37: error: fn is called with 1 arg but expects 0
src/status_im2/contexts/communities/discover/view.cljs:99:23: error: fn is called with 1 arg but expects 0
src/status_im2/contexts/onboarding/create_profile/view.cljs:76:69: error: fn is called with 1 arg but expects 0
src/status_im2/contexts/onboarding/create_profile/view.cljs:80:69: error: fn is called with 1 arg but expects 0
linting took 7357ms, errors: 5, warnings: 0
```
Might be due to Clojure upgrade, or Clj-kondo upgrade, no idea.

**Comment by jakubgs** [comment]
Also APK Analyzer fails to start:
```
 > make shell TARGET=android-sdk
Configuring Nix shell for target 'android-sdk'...

[nix-shell:~/work/status-mobile]$ apkanalyzer --version
Exception in thread "main" java.lang.IllegalStateException: Cannot locate latest build tools
	at com.android.tools.apk.analyzer.AaptInvoker.getPathToAapt(AaptInvoker.java:99)
	at com.android.tools.apk.analyzer.AaptInvoker.<init>(AaptInvoker.java:41)
	at com.android.tools.apk.analyzer.ApkAnalyzerCli.getAaptInvokerFromSdk(ApkAnalyzerCli.java:278)
	at com.android.tools.apk.analyzer.ApkAnalyzerCli.main(ApkAnalyzerCli.java:129)
```

**Comment by flexsurfer** [comment]
> Looks like linting is failing with:

сс @yqrashawn could you take a look pls

**Comment by yqrashawn** [comment]
guess these checks are from clj-kondo
add a commit to fix the lint

**Comment by flexsurfer** [comment]
thanks @yqrashawn , @jakubgs should be fixed now

**Comment by jakubgs** [comment]
There's two failure modes for `apkanalyzer`. One is an error about missing class `XmlSchema`:
```
Exception in thread "main" java.lang.NoClassDefFoundError: javax/xml/bind/annotation/XmlSchema
	at com.android.repository.api.SchemaModule$SchemaModuleVersion.<init>(SchemaModule.java:156)
	at com.android.repository.api.SchemaModule.<init>(SchemaModule.java:75)
	at com.android.sdklib.repository.AndroidSdkHandler.<clinit>(AndroidSdkHandler.java:81)
	at com.android.tools.apk.analyzer.ApkAnalyzerCli.getAaptInvokerFromSdk(ApkAnalyzerCli.java:277)
	at com.android.tools.apk.analyzer.ApkAnalyzerCli.main(ApkAnalyzerCli.java:129)
Caused by: java.lang.ClassNotFoundException: javax.xml.bind.annotation.XmlSchema
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	... 5 more
```
Which can be induced by changing from `openjdk8` to `openjdk11` in current `develop` branch.

The other one seems to be due to inability to find Android SDK `build-tools`:
```
Exception in thread "main" java.lang.IllegalStateException: Cannot locate latest build tools
	at com.android.tools.apk.analyzer.AaptInvoker.getPathToAapt(AaptInvoker.java:99)
	at com.android.tools.apk.analyzer.AaptInvoker.<init>(AaptInvoker.java:41)
	at com.android.tools.apk.analyzer.ApkAnalyzerCli.getAaptInvokerFromSdk(ApkAnalyzerCli.java:278)
	at com.android.tools.apk.analyzer.ApkAnalyzerCli.main(ApkAnalyzerCli.java:129)
```
The only obvious change are just Android SDK `platform-tool`, but they do not contain `apkanalyzer`.
And we can confirm the tool itself did not change because the JAR has the same hash:
```
[nix-shell:~/soft/status-mobile]$ git rev-parse --abbrev-ref HEAD
develop
[nix-shell:~/soft/status-mobile]$ md5sum $ANDROID_HOME/tools/lib/apkanalyzer-cli.jar
34c28934955fec95311d095340b791ff  /nix/store/wlmfw9yfrpcsw3dspddgalsgax4ahg67-androidsdk-mod-sdk/tools/lib/apkanalyzer-cli.jar
```
```
[nix-shell:~/work/status-mobile]$ git rev-parse --abbrev-ref HEAD
nixpkgs-upgrade
[nix-shell:~/work/status-mobile]$ md5sum $ANDROID_HOME/tools/lib/apkanalyzer-cli.jar
34c28934955fec95311d095340b791ff  /nix/store/lak09720695z7rn2mvha8d6ph2pn6x9g-androidsdk-mod-sdk/tools/lib/apkanalyzer-cli.jar
```
So the surrounding tooling, the bash wrapper around the tool, or something about Android SDK must have changed.

@yakimant I'd appreciate if you could investigate what's the issue here. The most likely fix will probably be somewhere in [`nix/pkgs/android-sdk/shell.nix`](https://github.com/status-im/status-mobile/blob/nixpkgs-upgrade/nix/pkgs/android-sdk/shell.nix), either in modifying `PATH`, or possibly in adjusting structure of the SDK distribution itself.

I have done some initial research into how `build-tools` is found, and here's the sources:
https://android.googlesource.com/platform/tools/base/+/studio-master-dev/apkparser/analyzer/src/main/java/com/android/tools/apk/analyzer/AaptInvoker.java#99
https://github.com/JetBrains/adt-tools-base/blob/e5faa7ab/sdklib/src/main/java/com/android/sdklib/repository/AndroidSdkHandler.java#L703-L735

**Comment by yakimant** [comment]
Summary of investigation:
- **SDK Tools** need **JDK8** and don't work with **JDK11** (`java.lang.NoClassDefFoundError: javax/xml/bind/annotation/XmlSchema` issue)
- **Command-line tools** need **JDK11** and don't work with **JDK8** (`java.lang.UnsupportedClassVersionError: com/android/prefs/AndroidLocationsProvider has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0` issue)
- **SDK Tools** (at least `apkanalyzer`) is broken in **nixpkgs** `androidenv` after upgrade. `build-tools` are not linked, but needed. (`java.lang.IllegalStateException: Cannot locate latest build tools` issue) [See in code](https://github.com/status-im/nixpkgs/blob/e7603eba51f2c7820c0a182c6bbb351181caa8e7/pkgs/development/mobile/androidenv/compose-android-packages.nix#L147-L156)

Tools versions
- **Command-line tools** are replacement for **SDK Tools**, which are deprecated and outdated ([2017](https://developer.android.com/tools/releases/sdk-tools))
- **JDK8** is still [maintained](https://en.wikipedia.org/wiki/Java_version_history) somehow

There are 2 possible solutions of upgrade issues:
1. Switch to **JDK11** and **Command-line tools**. Neet to test everything, bus a benefit we switch from outdated tooling.
2. Fix **nixpkgs** `androidenv` with PR, fork or override.

**Comment by jakubgs** [comment]
I think upgrading to OpenJDK 11 is a good idea. We should do it eventually, so it would be good to try it ~~in a separate PR~~.

Thanks for researching.

**Comment by yakimant** [comment]
We've switched:
- Java `1.8.0_322` to `11.0.11`
- **Android SDK tools** to **Android SDK Command-Line Tools** (higher in `PATH`)

As I see CI jobs pass succesfully, so we need some testing from Mobile developers.
@siddarthkay, @cammellos, can you please help to check, that everything is OK?

I also created an issue for broken `Android SDK Tools`: https://github.com/NixOS/nixpkgs/issues/237721

**Comment by siddarthkay** [comment]
> need some testing from Mobile developers

@status-im/mobile-qa

**Comment by siddarthkay** [comment]
@yakimant : I'll do some basic sanity checks on the builds but better to loop in QA on manual testing as well :)

**Comment by flexsurfer** [comment]
it's better to first rebase and fix conflicts

File: nix/pkgs/android-sdk/shell.nix
Comment (jakubgs) [inline nix/pkgs/android-sdk/shell.nix]: Hardcoding the same version is not great.

I thought we could get it from `compose` object, like you can for other sub-packages:
```
 > make nix-repl
nix repl default.nix
warning: future versions of Nix will require using `--file` to load a file
Welcome to Nix 2.14.1. Type :? for help.

Loading installable ''...
Added 4 variables.
nix-repl> pkgs.androidEnvCustom.compose.platform-tools.version
"33.0.3"

nix-repl> pkgs.androidEnvCustom.compose.ndk-bundle.version
"22.1.7171670"
```
I found this:
```
nix-repl> :p pkgs.androidEnvCustom.compose.androidsdk.drvAttrs.version
"9.0"
```
But I haven't checked if this is exactly the same value and can be trusted. You'd have to check the derivation.

We could also just use `echo` and a wildcard:
```sh
 > echo /tmp/test/*/bin | tr ' ' ':'
/tmp/test/1.1/bin:/tmp/test/1.2/bin
```

**Comment by yakimant** [comment]
Small update on why `apkanalyzer` from `Android SDK tools` on `develop` works with JDK11 upgrade from #16016.

It doesn't actually. It works with local java, not nixpkgs.
The following changes are missing:
https://github.com/status-im/status-mobile/blob/f1876c9c716edef3558c8cbd554190ad7efa67b8/nix/pkgs/android-sdk/shell.nix#L3-L7

So historically `android-sdk` shell was running with local Java.
This PR will fix it.

Another small finding:
https://github.com/status-im/status-mobile/blob/8fdfeb5e45b09a5ff9e1fd08fc86e30af226e01e/nix/shells.nix#L60
Probably we should always refer to `openjdk` as a single source of truth for JDK package used everywhere.

**Comment by yakimant** [comment]
Rebased from `develop` with fixes for `openjdk` mentioned above.

So again, I need some help with testing for:
1. JDK11 for `android-sdk` shell runs if you have any
2. Android SDK Tools switch

Assuming, that JDK11 works well for all other cases in `develop` - it's not much to check.

**Comment by jakubgs** [comment]
That's a good point. I did indeed add `buildInputs = [ openjdk ];` in `nix/pkgs/android-sdk/shell.nix` in my commit.

I guess that's the change that is the main difference from current state of `develop`.
Because I can confirm it does use system one:
```
[nix-shell:~/work/status-mobile]$ which java
/etc/profiles/per-user/jakubgs/bin/java
```

**Comment by jakubgs** [comment]
Lets squash this, rabase, make this non-draft, drop WIP in title, and get some reviews going.

Thanks for your help @yakimant .

**Comment by qoqobolo** [comment]
Hi, as soon as this PR receives at least one approval from the developers, please move it to the `E2E tests` column, at the same time we will take it for manual testing.
Thanks!

File: nix/overlay.nix
Comment (siddarthkay) [inline nix/overlay.nix:35]: are we getting rid of zprint?

**Review by siddarthkay — APPROVED** [review]

**Comment by churik** [comment]
@jakubgs 
tried to build nightly build via `manual` -  success.
PR builds can be installed - multiaccounts created, e2e tests are passed (failures are not related to PR, this is an issue with message delivery)

I guess PR can be merged, if @jakubgs has more ideas what can be checked here - please share.
Thank you!

**Pull request merged**
